app/controllers/application_controller.rb

```diff
@@ -20,28 +19,52 @@ def jsonapi_render(collection, filters)
   end
 
   def set_user
-    @current_user = if auth_token.present?
-        User.from_jwt(Authenticator.from_token(auth_token))
-      else
-        nil
-      end
+    return nil unless auth_token.present?
+
+    @current_user = User.where(
+                      id: client.authenticated_user_id,
+                      login: client.authenticated_user_login
+                    ).first_or_create do |user|
+                      user.display_name = client.authenticated_user_display_name
+
+                      # Explicitly set user admin accessor if encoded in JWT
+                      user.admin = client.authenticated_admin?
+                    end
+
+    if needs_roles_refresh?
+      set_roles
+    end
   end
 
   def set_roles
     return unless current_user
-    current_user.roles = panoptes_client.roles current_user.id
+
+    current_user.update(roles: panoptes.roles(current_user.id), roles_refreshed_at: Time.now)
   end
 
-  def panoptes_client
-    @panoptes_client ||= PanoptesClient.new auth_token
+  def panoptes
+    @panoptes_api ||= PanoptesApi.new auth_token
+  end
+
+  def client
```

Comment:
It feels like the `def panoptes` `def client`, `def token_created_at` methods should be moved to the `PanoptesApi` class as they are concerning the token handling / information from the API.

Reply:
the `panoptes` method is instantiating the PanoptesApi class for a controller instance, which can't really go in the PanoptesApi class. The `client` method is just a little shortcut to the gem's Panoptes::Client directly, which has to be broken out because I need access to my lib's methods and the Panoptes gem's `::Client` methods, as well as the gem's `::Client.panoptes` API-calling methods. 

`token_created_at` could probably move, though, yeah.

It's a bit confusing, which is why I tried to rename things.

Reply:
the wording & call sites seems conflated. If you are making a `PanoptesApi ` class to handle all the API comms, etc I would stick all the funcationality in there and I would also add delegator methods to the underlying clients you want to hit (i.e. expose them via the public interface of `PanoptesApi`) 

That way you've got one consistent interface from your app level controllers to the `PanoptesApi`

Reply:
So just remove the `client` method and move `token_created_at`? I was trying to avoid repeating `panoptes_api.client.`but maybe that's preferable.

Reply:
This and delegation cleaned it up, thanks.